app/prerender/service.ts

```diff
@@ -0,0 +1,42 @@
+import { A } from '@ember/array';
+import { assert } from '@ember/debug';
+import Service from '@ember/service';
+import { task, waitForQueue } from 'ember-concurrency';
+
+export default class Prerender extends Service {
+    pending = A();
+
+    ready = task(function* () {
+        yield waitForQueue('afterRender');
+        if (!this.get('pending.length')) {
+            window.prerenderReady = true;
+        }
+    }).drop();
+
+    waitOn(this: Prerender, key: string): void {
+        if (window.prerenderReady) {
+            return;
+        }
+        const pending = this.get('pending');
+        assert('Already waiting on that key', !pending.includes(key));
+        pending.pushObject(key);
+    }
+
+    finished(this: Prerender, key: string): void {
+        if (window.prerenderReady) {
+            return;
+        }
+        const pending = this.get('pending');
+        assert('Call `waitOn` before `finished`', pending.includes(key));
```

Comment:
Did you mean `!pending.includes(key)`?

Reply:
I'm thinking that when you call `finished`, the key should already be in `pending` from when you called `waitOn` earlier.

Reply:
Ignore me, I wasn't thinking clearly about how `assert()` works.